File: backend/app/models/retro.py
from datetime import datetime, timezone

from beanie import Document, PydanticObjectId
from pydantic import BaseModel, Field

REVEAL = "reveal"
CLUSTER = "cluster"
VOTE = "vote"
DISCUSS = "discuss"
DONE = "done"

# The one place the order lives. Transition checks are derived from it.
PHASE_ORDER = (REVEAL, CLUSTER, VOTE, DISCUSS, DONE)

# The vote budget one member gets, spent in a single submission (#8).
MAX_VOTES_PER_USER = 3

# Where a topic stands in the discussion (#9). The first entry is the default a
# freshly generated topic starts on.
TOPIC_STATUSES = ("pending", "discussed", "skipped", "deferred")

# An action item is open until somebody does it (#9).
ACTION_STATUSES = ("open", "done")

# Where a transcript extraction stands (#10). `idle` is never stored — it is what
# a retro with no `ai_suggestions` at all reads as — and the other three are the
# only values `ai_suggestions["status"]` ever holds. #17 polls this field.
EXTRACTION_STATUSES = ("idle", "processing", "ready", "failed")

# Where one AI suggestion stands. Nothing is deleted on confirm or reject, so the
# facilitator can see what became of every extracted line (#10).
SUGGESTION_STATES = ("pending", "confirmed", "rejected")

# About a two-hour meeting. A body-shape rule, so over it is 422 (#10).
MAX_TRANSCRIPT_CHARS = 100_000

# The hard ceiling on one proxy call. Without it a hung proxy would leave the
# retro stuck on `processing`, and a paste is refused while it is (#10).
PROXY_TIMEOUT_SECONDS = 60


def next_phase(current: str) -> str | None:
    """The only phase that may follow `current`, or None at the end."""
    index = PHASE_ORDER.index(current)
    if index + 1 >= len(PHASE_ORDER):
        return None
    return PHASE_ORDER[index + 1]


def _now() -> datetime:
    return datetime.now(timezone.utc)


class Cluster(BaseModel):
    id: str
    name: str
    created_at: datetime = Field(default_factory=_now)


class Vote(BaseModel):
    """One member's whole ballot, not one vote.

    Submission is atomic and cannot be retracted (`_docs/decisions.md`), so a
    member has at most one of these per retro. That is what makes "has Alice
    voted?" a presence check and a second submission a flat 409.

    Stacking is repetition: `["c1", "c1", "c1"]` is three votes on `c1`, and a
    cluster's count is how many times its id appears across every ballot.
    """

    user_id: PydanticObjectId
    cluster_ids: list[str]
    submitted_at: datetime = Field(default_factory=_now)


class Topic(BaseModel):
    """One cluster, promoted to something the team actually talks about (#9).

    Generated from `tally()` on entering `discuss`, never by hand. There is
    deliberately no `name`: the name is the cluster's, the cluster is in this
    same document, and cluster writes stopped when the `cluster` phase ended
    (#7) — so a stored copy could only duplicate a value that can no longer
    change. `app/services/discussion.py` resolves it into responses instead.

    `vote_count` and `rank` are a snapshot taken at generation, not a live view.
    Voting is over by then and there is no retraction (#8), so nothing can move
    them afterwards. Only `status` and `notes` are mutable.
    """

    id: str
    cluster_id: str
    vote_count: int = 0
    rank: int = 0
    status: str = TOPIC_STATUSES[0]
    notes: str = ""


class Decision(BaseModel):
    """Something the team settled on, optionally hung off a topic.

    `is_confirmed` is what separates an AI draft (#10) from a decision the
    facilitator stands behind; #11 publishes only the confirmed ones.
    """

    id: str
    topic_id: str | None = None
    text: str
    is_confirmed: bool = False


class Action(BaseModel):
    """A commitment with an owner, or without one until somebody takes it.

    `owner_name` belongs to #10's best-effort matching of AI-extracted owner
    strings. Nothing in #9 writes it or returns it — an action created through
    the discussion endpoints leaves it `None`.
    """

    id: str
    topic_id: str | None = None
    description: str
    owner_id: PydanticObjectId | None = None
    owner_name: str | None = None
    status: str = ACTION_STATUSES[0]
    due_date: datetime | None = None


class Retrospective(Document):
    cycle_id: PydanticObjectId
    phase: str = REVEAL
    clusters: list[Cluster] = Field(default_factory=list)
    votes: list[Vote] = Field(default_factory=list)
    topics: list[Topic] = Field(default_factory=list)
    decisions: list[Decision] = Field(default_factory=list)
    actions: list[Action] = Field(default_factory=list)
    # When the tally first became visible to anybody (#21). Set once, never
    # cleared and never moved: results visibility is monotonic, so a later
    # membership change cannot re-hide them or re-open withdrawal.
    voting_results_opened_at: datetime | None = None
    transcript: str | None = None
    ai_suggestions: dict | None = None
    created_at: datetime = Field(default_factory=_now)

    class Settings:
        name = "retrospectives"

File: frontend/src/api/retro.ts
/**
 * The retrospective endpoints, typed once (#16).
 *
 * `GET /api/retros/{id}` is authoritative for phase, clusters, who has
 * submitted a ballot, topics, decisions and actions. Cards come from the
 * cycle's feedback endpoint, and member identities from #31's dashboard — this
 * page never derives a display name or a facilitator from `created_by`.
 *
 * `votes` here is a list of submitters and their timestamps. It carries no
 * cluster ids, by design (#8), and nothing in this file asks for any.
 */
import { api } from "./client";
import type { FeedbackCard } from "./feedback";

export const PHASES = ["reveal", "cluster", "vote", "discuss"] as const;
export type Phase = (typeof PHASES)[number];
export const DONE = "done";

export const TOPIC_STATUSES = ["pending", "discussed", "skipped", "deferred"] as const;
export const MAX_VOTES = 3;

export type Cluster = { id: string; name: string; created_at: string };
export type VoteSubmitter = { user_id: string; submitted_at: string };
export type Topic = {
  id: string;
  cluster_id: string | null;
  name_override: string | null;
  name: string;
  vote_count: number;
  rank: number;
  status: string;
  notes: string;
};
export type Decision = {
  id: string;
  topic_id: string | null;
  text: string;
  is_confirmed: boolean;
};
export type Action = {
  id: string;
  topic_id: string | null;
  description: string;
  owner_id: string | null;
  owner_name: string | null;
  status: string;
  due_date: string | null;
};

export type Retro = {
  id: string;
  cycle_id: string;
  phase: string;
  clusters: Cluster[];
  votes: VoteSubmitter[];
  topics: Topic[];
  decisions: Decision[];
  actions: Action[];
  /** When the tally first became visible; null while it is still hidden (#21). */
  voting_results_opened_at: string | null;
  transcript: string | null;
  ai_suggestions: unknown;
  created_at: string;
};

export type VoteResults = {
  members_voted: number;
  members_total: number;
  total_votes: number;
  results: { cluster_id: string; name: string; vote_count: number; rank: number }[];
};

export type SuggestedGrouping = {
  clusters: { name: string; card_ids: string[] }[];
  ungrouped_card_ids: string[];
};

export async function getRetro(retroId: string): Promise<Retro> {
  return (await api.get<Retro>(`/api/retros/${retroId}`)).data;
}

export async function advancePhase(retroId: string, phase: Phase): Promise<Retro> {
  return (await api.patch<Retro>(`/api/retros/${retroId}/phase`, { phase })).data;
}

export async function createCluster(retroId: string, name: string): Promise<Cluster> {
  return (await api.post<Cluster>(`/api/retros/${retroId}/clusters`, { name })).data;
}

export async function renameCluster(
  retroId: string,
  clusterId: string,
  name: string
): Promise<Cluster> {
  return (await api.patch<Cluster>(`/api/retros/${retroId}/clusters/${clusterId}`, { name })).data;
}

export async function deleteCluster(retroId: string, clusterId: string): Promise<Cluster[]> {
  return (await api.delete<Cluster[]>(`/api/retros/${retroId}/clusters/${clusterId}`)).data;
}

export async function moveCard(cardId: string, clusterId: string | null): Promise<FeedbackCard> {
  return (await api.patch<FeedbackCard>(`/api/feedback/${cardId}/cluster`, {
    cluster_id: clusterId,
  })).data;
}

export async function suggestClusters(retroId: string): Promise<SuggestedGrouping> {
  return (await api.post<SuggestedGrouping>(`/api/retros/${retroId}/clusters/suggest`, {})).data;
}

/** One atomic ballot, repetitions preserved — stacked votes are repeated ids. */
export async function submitVotes(retroId: string, clusterIds: string[]): Promise<void> {
  await api.post(`/api/retros/${retroId}/votes`, { cluster_ids: clusterIds });
}

/** Withdraw the caller's whole ballot. There is no partial version (#21). */
export async function retractBallot(retroId: string): Promise<void> {
  await api.delete(`/api/retros/${retroId}/votes`);
}

export async function getResults(retroId: string): Promise<VoteResults> {
  return (await api.get<VoteResults>(`/api/retros/${retroId}/votes/results`)).data;
}

export async function createTopic(
  retroId: string,
  body: { name: string; rank?: number }
): Promise<Topic> {
  return (await api.post<Topic>(`/api/retros/${retroId}/topics`, body)).data;
}

export async function deleteTopic(retroId: string, topicId: string): Promise<void> {
  await api.delete(`/api/retros/${retroId}/topics/${topicId}`);
}

export async function updateTopic(
  retroId: string,
  topicId: string,
  // name 是覆盖,传 null 清掉就回到 cluster 的名字;rank 是 1 起的位置 (#22)
  body: { status?: string; notes?: string; name?: string | null; rank?: number }
): Promise<Topic> {
  return (await api.patch<Topic>(`/api/retros/${retroId}/topics/${topicId}`, body)).data;
}

export async function createDecision(
  retroId: string,
  body: { topic_id: string | null; text: string }
): Promise<Decision> {
  return (await api.post<Decision>(`/api/retros/${retroId}/decisions`, body)).data;
}

export async function updateDecision(
  retroId: string,
  decisionId: string,
  body: { text?: string; is_confirmed?: boolean; topic_id?: string | null }
): Promise<Decision> {
  return (await api.patch<Decision>(`/api/retros/${retroId}/decisions/${decisionId}`, body)).data;
}

export async function deleteDecision(retroId: string, decisionId: string): Promise<void> {
  await api.delete(`/api/retros/${retroId}/decisions/${decisionId}`);
}

export async function createAction(
  retroId: string,
  body: { topic_id: string | null; description: string; owner_id: string | null }
): Promise<Action> {
  return (await api.post<Action>(`/api/retros/${retroId}/actions`, body)).data;
}

export async function updateAction(
  retroId: string,
  actionId: string,
  body: Record<string, unknown>
): Promise<Action> {
  return (await api.patch<Action>(`/api/retros/${retroId}/actions/${actionId}`, body)).data;
}

export async function deleteAction(retroId: string, actionId: string): Promise<void> {
  await api.delete(`/api/retros/${retroId}/actions/${actionId}`);
}

/** The one legal next step, or null at `done`. Never a skip, never a reverse. */
export function nextPhase(phase: string): Phase | typeof DONE | null {
  const index = (PHASES as readonly string[]).indexOf(phase);
  if (index === -1) {
    return null;
  }
  return index + 1 < PHASES.length ? PHASES[index + 1] : DONE;
}

export function sortCards(cards: FeedbackCard[]): FeedbackCard[] {
  return [...cards].sort((left, right) => {
    if (left.created_at !== right.created_at) {
      return left.created_at < right.created_at ? -1 : 1;
    }
    return left.id < right.id ? -1 : 1;
  });
}
